File: house-robber.java
class Solution {
    public int rob(int[] nums) {
        int prevMax = 0;
        int currMax = 0;
        for (int num : nums) {
            // prevMax + num => take curr
            // currMax => do not take curr
            int temp = currMax;
            currMax = Math.max(prevMax + num, currMax);
            prevMax = temp;
        }
        
        return currMax;
    }
}

File: house-robber/house-robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        memo = [None] * len(nums)
        def helper(i):
            if i >= len(nums):
                return 0
            
            if memo[i] is not None:
                return memo[i]
            
            ans = max(nums[i] + helper(i + 2), helper(i + 1))
            memo[i] = ans
            return ans
        
        return helper(0)